### ISSUE TITLE
Delayed connection release?

### ISSUE THREAD

**Issue by andredasilvapinto** [AsyncHttpClient/async-http-client#1406]
I'm trying to limit the number of connections in use by blocking (instead of throwing exceptions) but it seems that the client keeps that the connection assignment is kept even after the request is finished?

I've created a simple case where I'm getting something that I would not expect: `Too many connections` errors.

Here is the snippet (just point it to some random port with no server behind it):

```
AsyncHttpClient client = new DefaultAsyncHttpClient(new Builder()
    .setConnectTimeout(1000)
    .setRequestTimeout(1000)
    .setMaxConnections(5)
    .build());

Semaphore semaphore = new Semaphore(5);

while (true) {
    semaphore.acquire();
    System.out.println("acquired");

    client.prepareGet("http://localhost:5000").execute(new AsyncHandler<Object>() {
        @Override
        public void onThrowable(Throwable t) {
            semaphore.release();
            System.out.println(t.getMessage() + ": released");
        }

        @Override
        public State onBodyPartReceived(HttpResponseBodyPart bodyPart) throws Exception {
            return State.CONTINUE;
        }

        @Override
        public State onStatusReceived(HttpResponseStatus responseStatus) throws Exception {
            return State.CONTINUE;
        }

        @Override
        public State onHeadersReceived(HttpResponseHeaders headers) throws Exception {
            return State.CONTINUE;
        }

        @Override
        public Object onCompleted() throws Exception {
            return null;
        }
    });
}
```

I was assuming that the semaphore would make sure the async client would never have more than 5 requests at a time and, consequently, never go beyond the maximum number of connections, but instead I'm getting this:

```
acquired
acquired
acquired
acquired
acquired
acquired
Request timeout to not-connected after 1000 ms: released
Too many connections: 5: released
acquired
Too many connections: 5: released
acquired
Too many connections: 5: released
acquired
Too many connections: 5: released
acquired
Too many connections: 5: released
```

Am I making the wrong assumption here? Why is the client not releasing the connection before calling `onThrowable`?

Version: 2.0.31

**Comment by andredasilvapinto** [comment]
I was debugging this and apparently for timeouts the client calls the error handler before releasing the connection `org.asynchttpclient.netty.channel.ChannelManager#freeChannels.release()`

**Comment by slandelle** [comment]
This is fixed in 2.1 (which is currently in alpha, latest is 2.1.0-alpha14).
Fix is complicated to backport in 2.0, so there's no plan to backport for now.

**Comment by andredasilvapinto** [comment]
Don't worry :) I didn't include the onComplete release in this scenario, because all of the request are failures and given that what I was trying to prove was that there were too many connections errors (and not too many blocked threads), having one single release makes it even clearer/simpler (so no possibility of having two releases for one single request/acquire).

Do you think there will be a stable 2.1 release in the near future? Meanwhile I think I will just remove the async http client max connections (set it to -1, the default) and rely on the external semaphore only.

Thanks for the reply!

**Comment by andredasilvapinto** [comment]
I can confirm that it is working as expected in 2.1.0-alpha12. I wasn't able to run the snippet with alpha13 or alpha14. I'm getting:

```
Exception in thread "main" java.lang.NoSuchMethodError: io.netty.handler.ssl.SslContextBuilder.protocols([Ljava/lang/String;)Lio/netty/handler/ssl/SslContextBuilder;
	at org.asynchttpclient.netty.ssl.DefaultSslEngineFactory.buildSslContext(DefaultSslEngineFactory.java:45)
	at org.asynchttpclient.netty.ssl.DefaultSslEngineFactory.init(DefaultSslEngineFactory.java:69)
	at org.asynchttpclient.netty.channel.ChannelManager.<init>(ChannelManager.java:110)
	at org.asynchttpclient.DefaultAsyncHttpClient.<init>(DefaultAsyncHttpClient.java:88)
	at Test.main(Test.java:17)
```

This happens because `io.netty.handler.ssl.SslContextBuilder.protocols()` was only introduced in netty 4.1.9.Final and because you have a version conflict for `netty-handler` with `com.typesafe.netty:netty-reactive-streams:jar:2.0.0-M1`, which uses 4.1.7.Final.

I see you specified the dependencyManagement entry, but you forgot to add the dependency for it here https://github.com/AsyncHttpClient/async-http-client/blob/master/client/pom.xml#L27-L56, therefore, when the client is used as a dependency, maven resolves the netty-handler version to the one that is explicitly used. It should be fixed if you add the dependency entry in the `async-http-client` module, which should always be the case, given that you are using it directly in there (you may consider running `mvn dependency:analyze` and adding maven enforcer to the build to detect these problems - they are not perfect, but they can help).

**Comment by slandelle** [comment]
> I see you specified the dependencyManagement entry, but you forgot to add the dependency for it here

Good catch. It's not that I forgot about it, but because I thought if was fine. Because dependencyManagement properly applies inside the project:

```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ async-http-client ---
[INFO] org.asynchttpclient:async-http-client:jar:2.1.0-SNAPSHOT
[INFO] +- org.asynchttpclient:async-http-client-netty-utils:jar:2.1.0-SNAPSHOT:compile
[INFO] |  \- io.netty:netty-buffer:jar:4.1.9.Final:compile
[INFO] +- io.netty:netty-codec-http:jar:4.1.9.Final:compile
[INFO] |  \- io.netty:netty-codec:jar:4.1.9.Final:compile
[INFO] +- io.netty:netty-transport-native-epoll:jar:linux-x86_64:4.1.9.Final:compile
[INFO] |  +- io.netty:netty-common:jar:4.1.9.Final:compile
[INFO] |  \- io.netty:netty-transport:jar:4.1.9.Final:compile
[INFO] +- io.netty:netty-resolver-dns:jar:4.1.9.Final:compile
[INFO] |  +- io.netty:netty-resolver:jar:4.1.9.Final:compile
[INFO] |  \- io.netty:netty-codec-dns:jar:4.1.9.Final:compile
[INFO] +- org.reactivestreams:reactive-streams:jar:1.0.0:compile
[INFO] +- com.typesafe.netty:netty-reactive-streams:jar:2.0.0-M1:compile
[INFO] |  \- io.netty:netty-handler:jar:4.1.9.Final:compile <=== indeed 4.1.9 here
```

But indeed maven doesn't seem to take it into account when another project resolves an AHC dependency. Another maven non sense...

**Comment by slandelle** [comment]
See #1409, thanks for reporting

> Do you think there will be a stable 2.1 release in the near future

Really can't say yet for sure. I might in deed release 2.1.0 so people get a "stable" version with Netty 4.1, then move to AHC 3 and start breaking things again...

**Comment by andredasilvapinto** [comment]
And by enforcing the netty-handler version on my project I get it to build but then I get a different error (both in alpha13 and alpha14):
```
Apr 28, 2017 9:25:32 PM io.netty.util.HashedWheelTimer$HashedWheelTimeout expire
WARNING: An exception was thrown by TimerTask.
java.lang.NullPointerException
	at org.asynchttpclient.netty.timeout.TimeoutTimerTask.appendRemoteAddress(TimeoutTimerTask.java:59)
	at org.asynchttpclient.netty.timeout.RequestTimeoutTimerTask.run(RequestTimeoutTimerTask.java:48)
	at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:661)
	at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:736)
	at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:464)
	at java.lang.Thread.run(Thread.java:745)
```

As the exception implies, `this.timeoutsHolder.remoteAddress()` returns null in this case. It is not being defined for the `TimeoutHolder` created during calls to `org.asynchttpclient.netty.request.NettyRequestSender#resolveAddresses`. Maybe you could use the `unresolvedRemoteAddress` that you pass on to the  `TimeoutHolder` constructor as `originalRemoteAddress`? Otherwise it's probably better to remove it from the signature as the argument is not being used anywhere in that class.

**Comment by slandelle** [comment]
That's #1407, whose fix is shipped in 2.1.0-alpha16. It's available on central, even thought it doesn't show up yet in the search engine.

**Comment by andredasilvapinto** [comment]
Indeed it is. Works great with that version + dependency override.
I'll be looking forward to the stable release. Thanks.

**Comment by aguptabbm** [comment]
@slandelle  I still facing below error with 2.1.0-alpha16  version.
org.asynchttpclient.exception.TooManyConnectionsException: Too many connections: 10
	at org.asynchttpclient.netty.channel.ConnectionSemaphore.acquireChannelLock(Unknown Source)

@andredasilvapinto, have you fixed this issue manually or with new version?

**Comment by andredasilvapinto** [comment]
@aguptabbm I removed the max connections configuration from async http client and I control its parallelism level just with my own semaphore, but I do this because I don't want to use an alpha version. Nevertheless, as I said above, I tested 2.1.0-alpha12 and I didn't experience the problem with the code I've posted on the initial message in this issue. Are you experiencing problems with the code I've posted above and 2.1.0-alpha16?
